src/extensions/default/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/PackageUpdaterStep.cs

```diff
@@ -55,126 +49,183 @@ public class PackageUpdaterStep : UpgradeStep
-        {
-            if (context is null)
-            {
-                throw new ArgumentNullException(nameof(context));
-            }
+            public override string Title => _analyzer.Name;
 
-            var project = context.CurrentProject.Required();
+            public override string Description => _analyzer.Name;
 
-            try
+            public override IEnumerable<UpgradeStep> SubSteps => _subSteps ?? Enumerable.Empty<UpgradeStep>();
+
+            protected override async Task<UpgradeStepInitializeResult> InitializeImplAsync(IUpgradeContext context, CancellationToken token)
             {
-                var projectFile = project.GetFile();
+                if (context is null)
+                {
+                    throw new ArgumentNullException(nameof(context));
+                }
 
-                var count = 0;
-                do
+                if (_subSteps is null)
                 {
-                    if (count >= MaxAnalysisIterations)
+                    var analysis = await GetAnalysisState(context, token).ConfigureAwait(false);
+
+                    if (analysis is null || !analysis.IsValid)
                     {
-                        Logger.LogError("Maximum package analysis and update iterations reached. Review NuGet dependencies manually");
-                        return new UpgradeStepApplyResult(UpgradeStepStatus.Failed, $"Maximum package analysis and update iterations reached");
+                        return new UpgradeStepInitializeResult(UpgradeStepStatus.Failed, "Package analysis failed", BuildBreakRisk.Unknown);
                     }
 
-                    if (_analysisState is not null)
+                    if (!analysis.AreChangesRecommended)
                     {
-                        projectFile.RemoveReferences(_analysisState.References.Deletions.Select(i => i.Item));
-
-                        projectFile.RemovePackages(_analysisState.Packages.Deletions.Select(i => i.Item));
-                        projectFile.AddPackages(_analysisState.Packages.Additions.Select(i => i.Item));
+                        Logger.LogInformation("No package updates needed");
+                        return new UpgradeStepInitializeResult(UpgradeStepStatus.Complete, "No package updates needed", BuildBreakRisk.None);
+                    }
 
-                        projectFile.RemoveFrameworkReferences(_analysisState.FrameworkReferences.Deletions.Select(i => i.Item));
-                        projectFile.AddFrameworkReferences(_analysisState.FrameworkReferences.Additions.Select(i => i.Item));
+                    var steps = new List<UpgradeStep>();
 
-                        await projectFile.SaveAsync(token).ConfigureAwait(false);
-                        count++;
+                    AddSubsteps(analysis.References.Deletions, "Remove reference '{0}'", static t => t.Name, static (file, op) => file.RemoveReferences(new[] { op.Item }));
```

Comment:
It might be simpler to just have this step make all the changes directly instead of having sub-steps for each change. I don't think users will typically need to add or remove packages one at a time instead of just applying all the changes from a given analyzer at once.

Reply:
It's about opting into the changes and knowing. This ends up looking like this:

![image](https://user-images.githubusercontent.com/583206/148830108-25160190-51ee-4a75-a07d-a6b90cd8d332.png)

There could be a number of changes the package maps end up making and this identifies what those are.

Reply:
Ok, this seems fine for now. I'm wondering if in cases where there are many packages that need added/removed we'll want to just have them all added/removed once per analyzer. Console output could list the changes that will be made during step initialization. I agree that's not as visible as this, though.

We can try this and change in the future if we find it's too many sub-steps.